### PR TITLE
 ContiguousBlockAllocator cleanup (code style, help)

### PR DESCRIPTION
**Pull request by jamshark70**
## Purpose and Motivation

A question arose on the forum about ContiguousBlockAllocator's `findNext` method where the answer wasn't evident in the code or the documentation.

Also, at the time I wrote this class, I favored "method call" syntax for `if` -- which I now find to obscure meaning. It had been in the back of my mind to clean up the code style for readability.

At the same time, it's a good opportunity to add comments explaining some of the implementation details, and expand the help file.

The only functional change is to deprecate `findNext` and `findPrevious` in favor of `pr*` versions (i.e. explicitly marked private). Users should not be calling these methods directly.

## Types of changes

- Documentation
- Breaking change (deprecation)

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

### CODE REVIEW

**Review by telephon — APPROVED** [review]
looks good to me.

**Comment by smrg-lm** [comment]
Apologies to insist, I still don't understand the index, it may be me thinking something wrong, I see:

ContiguousBlockAllocator prFindNext method receives 'addr' as an index to get the corresponding ContiguousBlock object or nil.

Then ContiguousBlockAllocator prFindNext search again in its own array but using the data from ContiguousBlock. Here I get an index out of range in sclang and python.

ContiguousBlock block start + size must be less than or equal to ContiguousBlockAllocator size - 1 because is a nested range.

By reading the code I understand that the index (although relative) is zero based and size is the amount of elements, so the the iteration index goes from start to start + size - 1. This rule should apply for both ContiguousBlockAllocator and ContiguousBlock start/offset and size. This is the only thing I find confusing and think it's a bug even if it doesn't fire in sclang.

**Comment by jamshark70** [comment]
> ContiguousBlockAllocator prFindNext method receives 'addr' as an index to get the corresponding ContiguousBlock object or nil.

Not exactly. The `start` of the returned ContiguousBlock must be > addr, *not* >= addr. If it's == address, then it's *this* block, not *next*.

> ContiguousBlock block start + size must be less than or equal to ContiguousBlockAllocator size - 1 because is a nested range.

This is not true. The last ContiguousBlock in the allocator: start + size will be equal to the allocator's size. If it's a 3-channel allocator, it starts off with a ContiguousBlock(0, 3, false).

So: If you have a sparse array where every a = ContiguousBlock(x, y) lives at `x` in the array, and the next block b starts at x + y, then the way to find the next block is to start with addr+1 and scan forward until finding non-nil or that end of the array.

```
i = addr + 1;
while { i < array.size and: { array[i].isNil } } {
    i = i + 1;
};
```

Now i is either the address of a ContiguousBlock or one past the end of the array. If it points to a block, then return `array[i]`. If it's past the end, return nil.

But in SC we can do this without a conditional because if i is out of bounds, `array[i]` returns nil. So *both* cases can be handled by simply `array[i]`.

- `if(i < array.size) { array[i] } { nil }`
- `array[i]`

Both return the same here.

But what if addr points to a block?

Then we already know where the next block is: "and the next block b starts at x + y."

So there is no need to scan. We can simply proceed as if i = block.start + block.size.

And from there, the above logic is the same. If this is within the array, then it points to a block and you can return it by array indexing. If it's beyond the array's bounds, then array indexing gives the right answer (nil).

If it would make you feel better, I could revise the code with the above `if`. But logically, it's redundant.

**Comment by smrg-lm** [comment]
I see, it is a bit disturbing for me because the logic of the algorithm is implicit in a very particular language feature, e.g. I couldn't think that an index out of range was an intended case and couldn't make sense of it.

**Comment by smrg-lm** [comment]
And now I'm smiling, that is why the offset didn't make sense in a note I took.

**Comment by telephon** [comment]
It's idiomatic (particular, but also very common).

E.g. you can write `x + (myArray[i] ? 0)` which will add nothing if out of bounds.

**Comment by jamshark70** [comment]
> Here I get an index out of range in sclang and python.

And:

> the logic of the algorithm is implicit in a very particular language feature

FWIW I think it's completely reasonable for code written in language X to use features that are unique to X. (One wouldn't insist that Haskell code not use argument matching dispatch because other languages don't have that feature.)

But I think this gives me the clue I needed as to the context behind this conversation -- now I remember that you're translating the SC class library into python.

So I'd now guess that the problem on your mind was "How can this method be written in Python when it steps outside array bounds and the purpose isn't clear from the code style?"

But the comment you made was "this method is written incorrectly and has a bug." 😮

The human analog to garbage in, garbage out: When the question is asked in an imprecise way, the answer may not be relevant to your purpose.

I think you can use the alternative in [my previous message](https://github.com/supercollider/supercollider/pull/5307#issuecomment-753580006), with the size check at the end -- though I realized you can avoid wasted iterations by `while { i <= top and: ... }` instead of `i < size`. `top` should always be the address of the last ContiguousBlock in the array -- all elements from `top+1` to the end should be nil, so there is no need to search them.

I think I'll add that to this PR. It's a clearer way to write it.

**Comment by smrg-lm** [comment]
> 
> But the comment you made was "this method seems to have a bug, doesn't it?" open_mouth
> 
> The human analog to garbage in, garbage out: When the question is asked in an imprecise way, the answer may not be relevant to your purpose.
> 

Yes, that's exactly what happened. Know that for me, in my mind, it was not possible then I asked the wrong question and of course you couldn't know either. Over the years I was learning about this elegant pitfalls but I couldn't sort this one and started to search for alternative things that make more sense in every other language and practices I know.

In general, I think that is not a good practice because makes it difficult to learn and to be aware of it. sclang is not a popular language with hundreds of books written about it, coding style was changing over the years and changes by developer, the only good (public) source are the forums and even there people start to take the concepts for granted and to be implicit about them.

Precisely these language things are the most criticized and generate rejection from people who want or have to learn SuperCollider, I know from experience, even people with previous knowledge in programming. It turns out to be not easy or weird, why to make it more complicated? Besides, the lack of reference concepts also generates uncertainty and that makes it more difficult.

**Comment by smrg-lm** [comment]
> E.g. you can write `x + (myArray[i] ? 0)` which will add nothing if out of bounds.

When I see that code I think: if the content of myArray at i is nil replace it by zero, that's easy but there is not way to know that the index can go out of range, it doesn't say: if the index goes out of range replace it by zero.

**Comment by smrg-lm** [comment]
> > E.g. you can write `x + (myArray[i] ? 0)` which will add nothing if out of bounds.
> 
> When I see that code I think: if the content of myArray at i is nil replace it by zero, that's easy but there is not way to know that the index can go out of range, it doesn't say: if the index goes out of range return zero.

That changes the meaning of the algorithm itself, moreover, to understand the written code someone has to know the algorithm it expresses.

**Comment by jamshark70** [comment]
I think a PR review should probably avoid higher level discussions of SC's deviance from conventions of other languages. They are fair questions, interesting ones, but this thread has a narrower scope.

I'll only suggest that perhaps there is no concrete reason why it should be a good thing to throw an array out of bounds exception. That it is conventional does not prove that it is good. Personally I find that kind of error to get in the way, without providing any useful information that couldn't be obtained another way -- the programmer has to waste time either avoiding or handling exceptions, instead of just doing the thing. But perhaps take this topic to the forum (as I'm partly playing devil's advocate here).

Summing up, then:

- The method as it was originally written was unnecessarily confusing.
- The latest update in this PR totally rewrites it, to make the meaning of the array index clear (`indexOfNextBlock`). Future readers will not have to wonder about the purpose of `temp.start + temp.size`.
- The final array indexing now has a comment explaining it. Well, there are two ways to address the out-of-bounds indexing:
  - A comment: zero performance impact but slight deviance from convention.
  - A conditional (`if(indexOfNextBlock < size)...`): Comforting, but the performance impact is not zero (by adding a comparison and a jump).
  For the core class library, I would favor more efficient execution.

**Comment by jamshark70** [comment]
@telephon stupid tablet clicked your name for review, and now I can't see where to cancel it. Feel free to ignore.

**Comment by dyfer** [comment]
@jamshark70 @telephone does this need further review or is it good to go?

**Comment by telephon** [comment]
I've looked through it all (right now I can't do more than that) and it reads good.

**Comment by brianlheim** [comment]
thanks @jamshark70 and @telephon !

@smrg-lm , James is absolutely right, this is not the right place for these discussions. Our forums and mailing lists would be a better place to discuss general feedback about the design of the language. Please avoid going off-topic in PR threads in the future, it's impolite to authors and reviewers.
